### PR TITLE
feat: remove varints

### PR DESCRIPTION
**Pull request by dignifiedquire**
and use `take` to have cleaner protocol reads

### CODE REVIEW

File: src/protocol.rs [hunk 1]
```diff
@@ -81,29 +81,31 @@ pub async fn write_lp<W: AsyncWrite + Unpin>(writer: &mut W, data: &[u8]) -> Res
     );
 
     // send length prefix
-    let mut buffer = [0u8; 10];
-    let lp = unsigned_varint::encode::u64(data.len() as u64, &mut buffer);
-    writer.write_all(lp).await?;
+    let data_len = data.len() as u64;
+    writer.write_u64_le(data_len).await?;
```
Comment (rklaehn) [inline src/protocol.rs:85]: not that it matters that much, but why le?

File: src/protocol.rs [hunk 2]
```diff
@@ -112,53 +114,41 @@ pub async fn read_lp<'a, R: AsyncRead + futures::io::AsyncRead + Unpin, T: Deser
...
+    debug!("finished reading");
+    Ok(buffer.split_to(size).freeze())
 }
 
 /// Return a buffer of the data, based on the length prefix, from the given source.
 /// The new buffer is split off from the buffer that is passed in the function.
-pub async fn read_lp_data<R: AsyncRead + futures::io::AsyncRead + Unpin>(
+pub async fn read_lp_data<R: AsyncRead + Unpin>(
     mut reader: R,
     buffer: &mut BytesMut,
```
Comment (rklaehn) [inline src/protocol.rs:142]: So do we still need this annoying buffer at all?

The issue I have is that when handing the remaining reader to the AsyncSliceDecoder, I have to use chain to take these bytes into account.

But when I do that I have no way to get the reader out again, meaning that I can only read one item...

**Review by rklaehn — APPROVED** [review]

File: src/protocol.rs (hunk 1 above)
Comment (flub) [inline src/protocol.rs:85]: Little-endian does offend the greybeard hiding in me a little.  Network byte-order has always been big-endian.  Though it is true that most mainline systems are little-endian these days.

Comment (dignifiedquire) [reply src/protocol.rs:85]: mostly a personal preference

File: src/protocol.rs (hunk 2 above)
Comment (dignifiedquire) [reply src/protocol.rs:142]: There should be no additional bytes in here anymore, see the `.take()` it now only reads the acutal required bytes, not more

File: src/protocol.rs (hunk 1 above)
Comment (rklaehn) [reply src/protocol.rs:85]: bao also uses le, so I guess it is consistent. Also I don't really mind.

File: src/protocol.rs [hunk 3]
```diff
@@ -81,29 +81,31 @@ pub async fn write_lp<W: AsyncWrite + Unpin>(writer: &mut W, data: &[u8]) -> Res
...
+    let size = usize::try_from(size)?;
+    let mut read = 0;
     while buffer.len() < size {
-        debug!("reading message, buffered {} of {size}", buffer.len());
-        if reader.read_buf(buffer).await? == 0 {
-            bail!("no more data available");
-        }
+        let r = reader.read_buf(buffer).await?;
+        ensure!(r != 0, "no more data available");
+        read += r;
```
Comment (flub) [inline src/protocol.rs]: Shouldn't one read to EOF if you're using `.take()`?

File: src/protocol.rs [hunk 4]
```diff
@@ -81,29 +81,31 @@ pub async fn write_lp<W: AsyncWrite + Unpin>(writer: &mut W, data: &[u8]) -> Res
...
     }
+    ensure!(size == read, "expected to read {} but read {}", size, read);
```
Comment (flub) [inline src/protocol.rs]: I wouldn't leave this in for prod code.  It really should be held up by the code just above already.

File: src/provider.rs [hunk 5]
```diff
@@ -312,7 +311,8 @@ async fn send_blob<W: AsyncWrite + Unpin>(
             debug!("writing data");
             let file = tokio::fs::File::open(&path).await?;
             let mut reader = tokio::io::BufReader::new(file);
-            tokio::io::copy(&mut reader, &mut writer).await?;
+            tokio::io::copy_buf(&mut reader, &mut writer).await?;
```
Comment (flub) [inline src/provider.rs:314]: lol, love the lack of docs on this...

File: src/protocol.rs (hunk 3 above)
Comment (dignifiedquire) [reply src/protocol.rs]: ups, yes

File: src/protocol.rs (hunk 1 above)
Comment (flub) [reply src/protocol.rs:85]: heh, i had the same comment.  In theory these days with most machines being LE there is less byte-swapping involved, which can be an argument for it (and I assumed it was, turns out it wasn't even that :smile: )

File: src/protocol.rs (hunk 4 above)
Comment (dignifiedquire) [reply src/protocol.rs]: fixed

File: src/protocol.rs (hunk 3 above)
Comment (dignifiedquire) [reply src/protocol.rs]: fixed

File: src/provider.rs (hunk 5 above)
Comment (dignifiedquire) [reply src/provider.rs:314]: yeah, so by reading the code, the TLDR is, use it if your sources provide AsyncBufRead, as that allows you to save on the additional buffer that `copy` creates.

File: src/protocol.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/protocol.rs:85]: 🤣

**Review by flub — APPROVED** [review]

**Pull request merged**